Minimap as HUD element with API control (#9079)
minetest/minetest: doc/lua_api.txt

* `offset`: offset in pixels from position.
* `alignment`: The alignment of the waypoint.

### `image_waypoint`

Same as `image`, but does not accept a `position`; the position is instead determined by `world_pos`, the world position of the waypoint.

* `scale`: The scale of the image, with 1 being the original texture size.
  Only the X coordinate scale is used (positive values).
  Negative values represent that percentage of the screen it
  should take; e.g. `x=-100` means 100% (width).
* `text`: The name of the texture that is displayed.
* `alignment`: The alignment of the image.
* `world_pos`: World position of the waypoint.
* `offset`: offset in pixels from position.

> 

Reviewer: Correct me if I'm wrong, but I think this implemented in a different PR. Seems to have been accidentally added to this PR while rebasing.

* `world_pos`: World position of the waypoint.
* `offset`: offset in pixels from position.

### `minimap`

Displays a minimap on the HUD.

* `size`: Size of the minimap to display. Minimap should be a square to avoid

>   anamorphosis.

Reviewer 2: ```suggestion
  distortion.
```

* `hud_set_hotbar_selected_image(texturename)`
    * sets image for selected item of hotbar
* `hud_get_hotbar_selected_image`: returns texturename
* `set_minimap_modes({modes}, selected_mode)`: Sets available minimap modes and
  selected mode number (existing modes are replaced). Modes is a table of mode
	descriptions. Mode description is a table with four fields :
    * `type`: Available type:
      * `off`: Minimap off
      * `surface`: Minimap in surface mode
      * `radar`: Minimap in radar mode
      * `texture`: Texture to be displayed instead of terrain map

>     * `label`: Label associated with this mode (displayed on minimap mode toggle)

Reviewer 3: As an alternative to `label` I'd suggest to use two translation strings:
`Minimap in surface mode, Zoom x%g` and
`Minimap in radar mode, Zoom x%g`.

The zoom factor can be calculated automatically and you don't need to worry about translating that stuff as soon the parser script can find those two somewhere in the code. I can see this param would give modders some more freedoms on naming the modes, but why would they need to be changed? The mode types are always the same - only the zoom may change.

Reviewer 4: This label is usefull for naming texture maps. I could make it optional, using your suggestions according to the mode. Problem is, how to determine zoom factor.
Default surface modes are : x1=256 nodes, x2=128 nodes, x4=64 nodes, and its double for radar modes. Imagine a modder wants to have, lets say, 512 nodes and 256 nodes for surface modes. What should be their zoom factor ?

Reviewer 3: I'd say that would be `0.5x` and `1x`. Taking 256 as surface reference.
It is indeed helpful to label texture maps, but not regular surface/radar maps. There it's common to have a string template to fill in the active zoom factor.

Reviewer 4: Ok, I'll take 256 as surface mode reference and 512 as radar mode reference for default labels. It'll be up to modders to do whatever they want if using custom sizes.

* `hud_set_hotbar_selected_image(texturename)`
    * sets image for selected item of hotbar
* `hud_get_hotbar_selected_image`: returns texturename
* `set_minimap_modes({modes}, selected_mode)`: Sets available minimap modes and
  selected mode number (existing modes are replaced). Modes is a table of mode
	descriptions. Mode description is a table with four fields :
    * `type`: Available type:
      * `off`: Minimap off
      * `surface`: Minimap in surface mode
      * `radar`: Minimap in radar mode

>       * `texture`: Minimap displays texture given in `texture` field instead of

Reviewer: The following is better, I think:

```
      * `texture`: Texture to be displayed instead of terrain map
```

Reviewer 4: Thanks !
Fixed also

        terrain map

>     * `label`: Label associated with this mode (displayed on F9)

Reviewer: F9 for minimap isn't a fixed keybind, and so using the name of the key isn't a good idea. Would it be better if the name of the action is used instead? e.g. `(displayed when pressing "Minimap" key)`

Reviewer 4: You're right ! Fixed :)

Reviewer: Thanks!

* `hud_set_hotbar_selected_image(texturename)`
    * sets image for selected item of hotbar
* `hud_get_hotbar_selected_image`: returns texturename
* `set_minimap_modes({modes}, selected_mode)`: Sets available minimap modes and
  selected mode number (existing modes are replaced). Modes is a table of mode

>   descriptions. Mode description is a table with four fields :

Reviewer 3: I'd keep this a bit shorter
```
* `set_minimap_modes({mode, mode, ...}, selected_mode)`
    * Overrides the available minimap modes and changes the selected mode.
      `mode` is a table consisting of up to four fields:
```

    * `type`: Available type:
      * `off`: Minimap off
      * `surface`: Minimap in surface mode
      * `radar`: Minimap in radar mode
      * `texture`: Texture to be displayed instead of terrain map (texture is
        centered around 0,0 and can be scaled)

>     * `label`: Label associated with this mode (displayed on minimap mode toggle)

Reviewer 3: ```suggestion
    * `label`: Optional label to display on minimap mode toggle
      The translation must be handled within the mod.
```

>     * `size`: Size of the terrain displayed in minimap (in map nodes)

Reviewer 3: ```suggestion
    * `size`: Size of the terrain displayed in minimap (in nodes)
```

    * `texture`: Only for texture type, name of the texture to display

>     * `scale`: Only for texture type, scale of the texture map in nodes per pixels

Reviewer 3: ```suggestion
    * `scale`: Only for texture type, scale of the texture map in nodes per pixel
```

       (for example a `scale` of 2 means each pixel represents a 2x2 nodes square)
  `selected_mode` is the mode index to be selected after modes have been changed

>   (0 is the fist mode).

Reviewer 3: ```suggestion
  (0 is the first mode).
```

* `hud_set_hotbar_selected_image(texturename)`
    * sets image for selected item of hotbar
* `hud_get_hotbar_selected_image`: returns texturename
* `set_minimap_modes({mode, mode, ...}, selected_mode)`

>     * Overrides the available minimap modes and changes the selected mode.

Reviewer 2: specify whether this allows customising the F9 toggle order

```suggestion
    * Overrides the available minimap modes (and toggle order), and changes the selected mode.
```

      `mode` is a table consisting of up to four fields:

>     * `type`: Available type:

Reviewer 2: all of the below needs extra level of indentation to indicate this is a child of the above

      * `off`: Minimap off
      * `surface`: Minimap in surface mode
      * `radar`: Minimap in radar mode
      * `texture`: Texture to be displayed instead of terrain map (texture is
        centered around 0,0 and can be scaled)
    * `label`: Optional label to display on minimap mode toggle
      The translation must be handled within the mod.

>     * `size`: Size of the terrain displayed in minimap (in nodes)

Reviewer 2: Diameter/Sidelength?

Reviewer 2: Mention 512 limit

* `hud_set_hotbar_selected_image(texturename)`
    * sets image for selected item of hotbar
* `hud_get_hotbar_selected_image`: returns texturename
* `set_minimap_modes({mode, mode, ...}, selected_mode)`
    * Overrides the available minimap modes (and toggle order), and changes the
	selected mode.
    * `mode` is a table consisting of up to four fields:
        * `type`: Available type:
            * `off`: Minimap off
            * `surface`: Minimap in surface mode
            * `radar`: Minimap in radar mode
            * `texture`: Texture to be displayed instead of terrain map (texture
            is centered around 0,0 and can be scaled). Texture size is limited

> 			to 512 x 512 pixel.

Reviewer 2: ```suggestion
            to 512 x 512 pixel.
```

        * `label`: Optional label to display on minimap mode toggle
          The translation must be handled within the mod.
        * `size`: Sidelength or diameter, in number of nodes, of the terrain

> 		displayed in minimap

Reviewer 2: ```suggestion
        displayed in minimap
```
